### PR TITLE
Migrated to vitest

### PR DESCRIPTION
**Pull request by mattpocock**
Fewer dependencies, more modern tooling, in the same spirit as #570

### CODE REVIEW

File: tests/index.test.ts [hunk 1]
```diff
@@ -518,37 +517,34 @@ describe('Utils', () => {
...
-          ok('Yooooo'),
-          ok(123),
-          ok(true),
+        type HeterogenousList = [
+          Result<string, string>,
+          Result<number, number>,
+          Result<boolean, boolean>,
         ]
 
-        type ExpecteResult = Result<[ string, number, boolean ], (string | number | boolean)[]>
```
Comment (mattpocock) [inline tests/index.test.ts:529]: Looks like prettier formatting

File: tests/tsconfig.tests.json [hunk 2]
```diff
@@ -1,26 +1,27 @@
...
+    "module": "ES2015",
+    "noImplicitAny": true,
+    "sourceMap": false,
+    "downlevelIteration": true,
+    "noUnusedLocals": false,
+    "noUnusedParameters": false,
+    "strictNullChecks": true,
+    "strictFunctionTypes": true,
+    "declaration": true,
+    "skipLibCheck": true,
```
Comment (mattpocock) [inline tests/tsconfig.tests.json:13]: Only change here is adding skipLibCheck: true to remove an erroneous error when I upgraded

**Review by dmmulroy — APPROVED** [review]
lgtm - thanks Matt

@m-shaka @supermacro I am much in favor of this change and have been using `vitest` by default for a year or so now

File: tests/tsconfig.tests.json (hunk 2 above)
Comment (supermacro) [reply tests/tsconfig.tests.json:13]: I was curious to know what the issues were. Here they are (at least for posterity): 

image

... interestingly `WeakKey` is infact a ts builtin:

image

As for the "Cannot find module 'rollup/parseAst' or its corresponding type declarations." issue, I came across this thread:

https://github.com/rollup/rollup/issues/5199

Any thoughts @dmmulroy

**Comment by supermacro** [comment]
Love this. Thank you 🙌

**Comment by mattpocock** [comment]
@supermacro Getting the same error as [here](https://github.com/supermacro/neverthrow/pull/574#issuecomment-2334663906) on this PR

File: tests/tsconfig.tests.json (hunk 2 above)
Comment (m-shaka) [reply tests/tsconfig.tests.json:13]: https://github.com/vitest-dev/vitest/issues/4921

`WeakKey` is available in TS >= 5.2

Comment (mattpocock) [reply tests/tsconfig.tests.json:13]: Errors thrown in `node_modules` don't matter - it's not code that we control. If there are no errors on install (i.e. peerDeps are all good) then we're happy.

**Comment by dmmulroy** [comment]
@mattpocock @m-shaka Can we get tests/CI passing and get this in?

**Review by m-shaka — APPROVED** [review]

**Comment by supermacro** [comment]
Too many conflicts, easier to just create a new branch. See #597

**Pull request closed without merging**